src/main/scala/Engine.scala

```diff
@@ -41,8 +41,9 @@ case class Query(
     returnSelf: Option[Boolean] = None,// means for an item query should the item itself be returned, defaults
                                        // to what is in the algorithm params or false
     num: Option[Int] = None, // default: whatever is in algorithm params, which itself has a default--probably 20
-    eventNames: Option[List[String]]) // names used to ID all user actions
-  extends Serializable
+    eventNames: Option[List[String]], // names used to ID all user actions
+    addRank: Option[Boolean] = None // Add to ItemScore rank fields values
+) extends Serializable
 
```

Comment:
No addRank wrong and is not necessary here, as I proposed the `backFillField` becomes `backfills` and is an array, whose order defines the precedence of the 3 ranking methods: popularity, user defined, and random.

The JSON in my comment was meant to be a new param and makes `addRank` redundant.

instead of `addRank` the param should be:

```
"backfills":[
     {
         "name": "popRank", // defines the property name in Elasticsearch
         "type": "popular",
         "eventNames": ["buy", "view"],
         "duration": "3 days", // note that this has changed from v0.2.3
         "endDate": "ISO8601-date" //most recent date to end the duration
    },{
         "name": "userDefined", // defines the property name in Elasticsearch
         "type": "userDefined",
    },{
         "name": "defaultRank",  // defines the property name in Elasticsearch
         "type": "random",
   }
]
```

Where the order is significant and defines precedence.